asmrun/amd64.S

```diff
@@ -122,95 +114,88 @@
-
-/* Switch from C to OCaml stack.  Clobbers REG. */
-#define SWITCH_C_TO_OCAML_NO_CTXT(REG,CFA_OFF) \
-    /* Switch to OCaml stack */ \
-        movq	Caml_state(current_stack), REG; \
-    /* REG is Stack_sp(caml_current_stack) */ \
-        movq    Stack_sp(REG), %rsp; \
-        .cfi_def_cfa_offset CFA_OFF
+        PUSH_EXN_HANDLER; \
+        SWITCH_OCAML_TO_C_NO_CTXT((PUSHED) + 16)
+
+/* Switch from C to OCaml stack.  Clobbers %r11. */
+#define SWITCH_C_TO_OCAML_NO_CTXT \
+    /* Assert that %rsp == Caml_state->c_stack &&
+        Caml_state->c_stack->sp == Caml_state->current_stack->sp */ \
+        IF_DEBUG(cmpq %rsp, Caml_state(c_stack); je 8f; int3; 8:\
+                 movq Caml_state(current_stack), %r11; movq Stack_sp(%r11), %r11; \
+                 cmpq %r11, Cstack_sp(%rsp); je 8f; int3; 8:) \
+        movq    Cstack_sp(%rsp), %rsp; \
+        .cfi_restore_state
 
 /* Switch from C to OCaml stack. Also pops the context
- * from the bottom of the OCaml stack. Clobbers REG. */
-#define SWITCH_C_TO_OCAML(REG) \
-        SWITCH_C_TO_OCAML_NO_CTXT(REG,24); \
+ * from the bottom of the OCaml stack. Clobbers %r11. */
+#define SWITCH_C_TO_OCAML \
+        SWITCH_C_TO_OCAML_NO_CTXT; \
     /* Pop the caml_context from the bottom of stack */ \
-        POP_EXN_HANDLER(REG); \
-        popq    REG; CFI_ADJUST(-8); \
+        POP_EXN_HANDLER; \
+        popq    %r11; CFI_ADJUST(-8); \
 
 /* Load Caml_state->exn_handler into %rsp and restores prior exn_handler. Clobbers %r11. */
 #define RESTORE_EXN_HANDLER_OCAML \
         movq	Caml_state(exn_handler), %rsp; \
         .cfi_def_cfa_offset 16; \
-        POP_EXN_HANDLER(%r11)
+        POP_EXN_HANDLER
 
-/* Switch between OCaml stacks.
- * arguments : target stack (%rdi)
- * preserves : %rbx, %r8, %r9, %r12, %r13 */
+/* Switch between OCaml stacks. Clobbers %r11. Expects target stack in %r10. */
 #define SWITCH_OCAML_STACKS \
-    /* Switch from OCaml to C */ \
-        SWITCH_OCAML_TO_C(%r10); \
-        ENTER_FUNCTION; \
-    /* Call switch stack */ \
-        PREPARE_FOR_C_CALL; \
-        movq    $0, %rsi; \
-        call    GCALL(caml_switch_stack); \
-        CLEANUP_AFTER_C_CALL; \
-    /* Load ocaml stack and restore global variables */ \
-        LEAVE_FUNCTION; \
-        SWITCH_C_TO_OCAML(%r10)
-
-/* Switch between OCaml stacks and free the current one.
- * arguments : target stack (%rdi)
- * preserves : %rbx, %r12, %r13 */
-#define SWITCH_OCAML_STACKS_AND_FREE \
-    /* Switch from OCaml to C */ \
-        SWITCH_OCAML_TO_C(%r10); \
-        ENTER_FUNCTION; \
-    /* Call switch stack */ \
-        PREPARE_FOR_C_CALL; \
-        movq    $1, %rsi; \
-        call    GCALL(caml_switch_stack); \
-        CLEANUP_AFTER_C_CALL; \
-    /* Load ocaml stack and restore global variables */ \
-        LEAVE_FUNCTION; \
-        SWITCH_C_TO_OCAML(%r10)
-
+    /* Build caml_context at the bottom of the stack */ \
+        pushq   $0 ; CFI_ADJUST(8); \
+        PUSH_EXN_HANDLER; \
```

Comment:
Perhaps useful to add a comment here that we're pushing an exception handler just to save the exception pointer in the source stack.

Reply:
Done